components/backup-stream/src/subscription_track.rs

```diff
@@ -0,0 +1,313 @@
+
+        let old_epoch = subscription.meta.get_region_epoch();
+        let new_epoch = new_region.get_region_epoch();
+        if old_epoch.version == new_epoch.version {
+            subscription.meta = new_region.clone();
+            return true;
+        }
+
+        false
+    }
+
+    /// check whether the region_id should be observed by this observer.
+    pub fn is_observing(&self, region_id: u64) -> bool {
+        let mut exists = false;
+
+        // The region traced, check it whether is still be observing,
+        // if not, remove it.
+        let still_observing = self
+            .0
+            // Assuming this closure would be called iff the key exists.
+            // So we can elide a `contains` check.
+            .remove_if(&region_id, |_, o| {
+                exists = true;
+                !o.is_observing()
+            })
+            .is_none();
+        exists && still_observing
+    }
+
+    pub fn get_subscription_of(&self, region_id: u64) -> Option<RefMut<u64, RegionSubscription>> {
+        self.0.get_mut(&region_id)
+    }
+}
+
+/// This enhanced version of `Resolver` allow some unorder of lock events.  
+/// The name "2-phase" means this is used for 2 *concurrency* phases of observing a region:
+/// 1. Doing the initial scanning.
+/// 2. Listening at the incremental data.
+///
+/// ```text
+/// +->(Start TS Of Task)            +->(Task registered to KV)
+/// +--------------------------------+------------------------>
+/// ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^ ^~~~~~~~~~~~~~~~~~~~~~~~~
+/// |                                 +-> Phase 2: Listening incremtnal data.
+/// +-> Phase 1: Initial scanning scans writes between start ts and now.
+/// ```
+///
+/// In backup-stream, we execute these two tasks parallelly. Which may make some race conditions:
+/// - When doing initial scanning, there may be a flush triggered, but the defult resolver
+///   would probably resolved to the tip of incremental events.
+/// - When doing initial scanning, we meet and track a lock already meet by the incremental events,
+///   then the default resolver cannot untrack this lock any more.
+///
+/// This version of resolver did some change for solve these problmes:
+/// - The resolver won't advance the resolved ts to greater than `stable_ts` if there is some. This
+///   can help us prevent resolved ts from advancing when initial scanning hasn't finished yet.
+/// - When we `untrack` a lock haven't been tracked, this would record it, and skip this lock if we want to track it then.
+///   This would be safe because:
+///   - untracking a lock not be tracked is no-op for now.
+///   - tracking a lock have already being untracked (unordered call of `track` and `untrack`) wouldn't happen at phase 2 for same region.
+///     but only when phase 1 and phase 2 happend concurrently, at that time, we wouldn't and cannot advance the resolved ts.
+pub struct TwoPhaseResolver {
+    resolver: Resolver,
+    future_locks: Vec<FutureLock>,
```

Comment:
if we use vector here. how to guarantee the `untrack` must appear before `track`.
can we handle the `untrack` happen from raft cmd and the `track` happen from initial scan?🤔

Reply:
OK， seems no problem.
we don't put lock in vec from initial sca.